### PR TITLE
Add GetType() method on all resources

### PR DESCRIPTION
**Pull request by matthchr**
This will be used by extension resources in the future.

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

**Review by theunrepentantgeek — APPROVED** [review]
Looks good. I noted a couple of opportunities to simplify code using the `astbuilder` package.

File: hack/generator/pkg/interfaces/kubernetes_resource_interface.go [hunk 1]
```diff
@@ -353,6 +359,34 @@ func createGetStatusFunction(f *functions.ObjectFunction, codeGenerationContext
...
+				X: receiver.AsType(codeGenerationContext),
+			},
+			Params: nil,
+			Body: []dst.Stmt{
+				&dst.ReturnStmt{
+					Results: []dst.Expr{
+						astbuilder.TextLiteral(armType),
+					},
+				},
+			},
```
[the remark this answers is not in the document]
Comment (matthchr) [reply hack/generator/pkg/interfaces/kubernetes_resource_interface.go:380]: Good catch. This was copied and adapted from one of the other functions in this file, which isn't using them (possibly written before they existed?)

[the remark this answers is not in the document]
Comment (matthchr) [reply hack/generator/pkg/interfaces/kubernetes_resource_interface.go:380]: Fixed

[the remark this answers is not in the document]
Comment (matthchr) [reply hack/generator/pkg/interfaces/kubernetes_resource_interface.go:380]: I've also made a note to in a clean-up PR clean up the other places in this file that are doing it without `astbuilder`, but will do that in a separate PR

File: hack/generator/pkg/codegen/pipeline/apply_kubernetes_resource_interface.go [hunk 2]
```diff
@@ -56,3 +61,44 @@ func ApplyKubernetesResourceInterface(idFactory astmodel.IdentifierFactory) Stag
...
+	if !ok {
+		return "", errors.Errorf("expected %s.Type to be of type astmodel.TypeName", armTypeDefinition.Name())
+	}
+
+	t, ok := types[typePropertyTypeName]
+	if !ok {
+		return "", errors.Errorf("couldn't find type %q", typePropertyTypeName)
+	}
+
+	enumType, ok := t.Type().(*astmodel.EnumType)
```
[the remark this answers is not in the document]
Comment (matthchr) [reply hack/generator/pkg/codegen/pipeline/apply_kubernetes_resource_interface.go:98]: Obviously _today_ this doesn't happen, but I don't see any issue with being defensive here, so will change this.

**Pull request merged**